### PR TITLE
add coerce option for component props

### PR DESCRIPTION
**Pull request by lepture**
Add a force convert option for component. Here is an example:

```html
<!-- Component Avatar -->
<script>
export default {
  props: {
    size: {
      type: Number,
      coerce: 'number',
      default: 48,
    }
  }
}
</script>
```

And then I can use:

```html
<avatar size="38"></avatar>
```

Otherwise, size will always be String, it won't pass the validation.

### CODE REVIEW

**Comment by lepture** [comment]
I will add tests and documentation after your acceptance.

File: src/compiler/compile-props.js [hunk 1]
```diff
@@ -220,3 +220,31 @@ function getDefault (vm, options) {
...
+ *
+ * @param {*} value
+ * @param {Object} options
+ * @return {*}
+ */
+
+function getCoercedValue (value, options) {
+  var coerce = options.coerce
+  if (coerce) {
+    return value
```
Comment (Pumpuli) [inline src/compiler/compile-props.js]: So if `options.coerce` exists the function does nothing...?

**Comment by yyx990803** [comment]
1. Instead of specifying coerce type separately, how about just `coerce: true`? Since the type is declared already. For example:

``` js
{
  type: Number,
  coerce: true
}
```

Will simply coerce the value into number.

2. In addition to initialization, the coercion should also happen when prop is updated: https://github.com/vuejs/vue/blob/dev/src/directives/internal/prop.js#L28-L30

**Comment by lepture** [comment]
@yyx990803 coerce can be a function.

File: src/compiler/compile-props.js (hunk 1 above)
Comment (lepture) [inline src/compiler/compile-props.js]: if (! coerce)

**Comment by lepture** [comment]
@yyx990803 Maybe it is better to coerce value in assertProp function.

**Comment by yyx990803** [comment]
@lepture yeah, coerce in `assertProp` sounds better.

For the value of the option, `number` and `int` sounds pretty arbitrary for me. How about `boolean` and `string`? It seems the following would be more reasonable:

1. If `coerce` is a function, use that function to coerce the value;
2. In other cases, `coerce` must be used together when `type` is also present, and it will coerce the value to the type specified. (there are internal util functions for these)
3. Special needs like `int` should just use a custom coerce function.

**Comment by lepture** [comment]
How about coerce accepts only function by now? We could implement other values later.

BTW, Boolean, String, Number are valid functions too.

**Comment by lepture** [comment]
Actually it should be in initProp

**Comment by yyx990803** [comment]
initProp doesn't affect when the prop is updated from the parent though

**Comment by lepture** [comment]
Yes. But it should be in initProp, since we have to assign the coerced value to vm. I'll parse the value in update bind.

**Comment by yyx990803** [comment]
We can extract `coerceProp` as a function into util, so that it can be used in both places like `assertProp` and `initProp`.

**Comment by lepture** [comment]
@yyx990803 updated. Should I export `coerceProp`?

**Comment by yyx990803** [comment]
Yeah, we should export it, and then call it before `assertProp` here: https://github.com/vuejs/vue/blob/dev/src/directives/internal/prop.js#L28

**Comment by lepture** [comment]
@yyx990803 updated. It should be documented and tested now.

File: src/directives/internal/prop.js [hunk 2]
```diff
@@ -25,6 +25,7 @@ export default {
       parent,
       parentKey,
       function (val) {
+        val = coerceProp(val)
```
Comment (yyx990803) [inline src/directives/internal/prop.js]: should be `coerceProp(prop, val)`

Comment (lepture) [inline src/directives/internal/prop.js]: OMG.

**Comment by yyx990803** [comment]
Cool it passes now, can you add a test case for it?

**Comment by lepture** [comment]
@yyx990803 updated with tests.

**Comment by yyx990803** [comment]
:+1:  Thanks!

**Comment by bekite** [comment]
Is it possible to access the vue instance inside the coerce function.
I have defined some helper methods I want to use.

```
    someObj: {
      coerce (val) {
        // return this.$helper.copy(val) <-- would be nice
        return JSON.parse(JSON.stringify(val))
      }
    },

**Pull request merged**